Userland/Libraries/LibJS/Runtime/PromiseConstructor.cpp

```diff
@@ -6,15 +6,131 @@
 
 namespace JS {
 
+// 27.2.4.1.1 GetPromiseResolve ( promiseConstructor ), https://tc39.es/ecma262/#sec-getpromiseresolve
+static Value get_promise_resolve(GlobalObject& global_object, Value constructor)
+{
+    VERIFY(constructor.is_constructor());
+    auto& vm = global_object.vm();
+
+    auto promise_resolve = constructor.get(global_object, vm.names.resolve);
+    if (vm.exception())
+        return {};
+    if (!promise_resolve.is_function()) {
+        vm.throw_exception<TypeError>(global_object, ErrorType::NotAFunction, promise_resolve.to_string_without_side_effects());
+        return {};
+    }
+
+    return promise_resolve;
+}
+
+// 27.2.1.1.1 IfAbruptRejectPromise ( value, capability ), https://tc39.es/ecma262/#sec-ifabruptrejectpromise
+static Optional<Value> if_abrupt_reject_promise(GlobalObject& global_object, Value, PromiseCapability const& capability)
+{
+    auto& vm = global_object.vm();
+
+    if (auto* exception = vm.exception()) {
+        vm.clear_exception();
+        vm.stop_unwind();
+
+        (void)vm.call(*capability.reject, js_undefined(), exception->value());
+        return capability.promise;
+    }
+
+    return {};
+}
+
+static bool iterator_record_is_complete(GlobalObject& global_object, Object& iterator_record)
+{
+    auto& vm = global_object.vm();
+
+    // FIXME: Create a native iterator structure with the [[Done]] internal slot. For now, temporarily clear
+    //        the exception so we can access the "done" property on the iterator object.
+    TemporaryClearException clear_exception(vm);
+    return iterator_complete(global_object, iterator_record);
+}
+
+static void set_iterator_record_complete(GlobalObject& global_object, Object& iterator_record)
+{
+    auto& vm = global_object.vm();
+
+    // FIXME: Create a native iterator structure with the [[Done]] internal slot. For now, temporarily clear
+    //        the exception so we can access the "done" property on the iterator object.
+    TemporaryClearException clear_exception(vm);
+    iterator_record.set(vm.names.done, Value(true), Object::ShouldThrowExceptions::No);
+}
+
+// 27.2.4.1.2 PerformPromiseAll ( iteratorRecord, constructor, resultCapability, promiseResolve ), https://tc39.es/ecma262/#sec-performpromiseall
+static Value perform_promise_all(GlobalObject& global_object, Object& iterator_record, Value constructor, PromiseCapability result_capability, Value promise_resolve)
+{
+    auto& vm = global_object.vm();
+
+    VERIFY(constructor.is_constructor());
+    VERIFY(promise_resolve.is_function());
+
+    auto* values = vm.heap().allocate_without_global_object<PromiseValueList>();
+    auto* remaining_elements_count = vm.heap().allocate_without_global_object<RemainingElements>(1);
+    size_t index = 0;
+
+    while (true) {
+        auto* next = iterator_step(global_object, iterator_record);
+        if (vm.exception()) {
+            set_iterator_record_complete(global_object, iterator_record);
+            return next;
```

Comment:
This doesn't look correct. In the exception case we get a nullptr for `next` - what even is the meaning of `Value` with an `Object*` nullptr (or any for that matter)? That sounds like something that shouldn't exist. Either way, I think we want an empty value here.

Reply:
Yeah, maybe I interpreted the spec too literally here:

https://tc39.es/ecma262/#sec-returnifabrupt
`If argument is an abrupt completion, return argument.`

Returning `{}` will be fine here in any case (and the same behavior), because a null `Object*` becomes a null `Value` anyways.

Reply:
>  If argument is an abrupt completion, return argument.

yeah, since they pass exceptions around as completion records, this just means "if there's an exception, don't clear it"

> because a null Object* becomes a null Value anyways.

Oh, I see: `m_type(object ? Type::Object : Type::Null)`
Not sure if I like that, but that's another story for another PR :sweat_smile:

> and the same behavior

Not exactly, empty value and null value are different. Doesn't matter though since it's discarded anyway.